### PR TITLE
fix: Delete dependent documents on deleteTeacher

### PR DESCRIPTION
**Pull request by carissa-tang**
## Notion ticket link

[Delete Test Sessions associated with teacher on deleteTeacher](https://www.notion.so/uwblueprintexecs/Delete-Test-Sessions-associated-with-teacher-on-deleteTeacher-38743ea27b1648cd9ea36e7df5d6bf31?pvs=4)
[Delete teacher id from associated School Object on deleteTeacher](https://www.notion.so/uwblueprintexecs/Delete-teacher-id-from-associated-School-Object-on-deleteTeacher-f6595a11dd6f46fcba0c250da9f203ab?pvs=4)

## Implementation description
* Updated `UserModel` middleware to do the following after `findOneAndDelete` is executed
   * Delete all test sessions containing the deleted teacher's id
   * Delete the teacher id from the School document which contains it
* Wrote unit tests for success and failure

## Steps to test
1. yarn test

## What should reviewers focus on?
* Info on middleware: https://mongoosejs.com/docs/middleware.html#use-cases

## Checklist
- [ ] My PR name is descriptive and in imperative tense
- [ ] My commit messages are descriptive and in imperative tense. My commits are atomic and trivial commits are squashed or fixup'd into non-trivial commits
- [ ] I have run the appropriate linter(s)
- [ ] I have requested a review from the PL, as well as other devs who have background knowledge on this PR or who will be building on top of this PR

### CODE REVIEW

File: backend/services/implementations/__tests__/userService.test.ts [hunk 1]
```diff
@@ -94,26 +102,16 @@ describe("mongo userService", (): void => {
   it("getAllTeachers", async () => {
     const createRes = await UserModel.insertMany(testUsers);
 
-    const testSchools = [
```
Comment (carissa-tang) [inline backend/services/implementations/__tests__/userService.test.ts:97]: refactor to use existing mock data

File: backend/models/user.model.ts [hunk 2]
```diff
@@ -62,4 +64,16 @@ const UserSchema: Schema = new Schema({
...
+  if (doc.role !== "Teacher") return;
+
+  /* eslint-disable no-underscore-dangle */
+  await MgTestSession.deleteMany({ teacher: doc._id });
+  await MgSchool.findOneAndUpdate(
+    { teachers: doc._id },
+    { $pull: { teachers: doc._id } },
+    { new: true },
+  );
+});
```
Comment (joyce-shi) [inline backend/models/user.model.ts:80]: bit of a nit: but should we use `pre` hook here instead of `post`? i.e delete child dependencies first and then the parent?

**Review by joyce-shi — APPROVED** [review]
lgtm! left one minor q

File: backend/models/user.model.ts (hunk 2 above)
Comment (carissa-tang) [reply backend/models/user.model.ts:80]: hmm that makes sense - but i was also thinking if user deletion fails, would we need to roll back the changes?

Comment (joyce-shi) [reply backend/models/user.model.ts:80]: LOL I was actually also thinking about that when I added this comment, i.e if deletion fails, is it better to have undeleted test sessions or undeleted teacher? 👀

i.e if you deleted the test session and it worked, but then the teacher didn't delete, it would still be fine (b/c teacher would just have one less test session). but if deleted teacher & the test session didn't delete, then there would be a test session w/ an undefined teacher field... and not sure if we handle that as gracefully? but maybe our code should just handle all those cases gracefully? idk

TLDR i'm not actually sure what is best here 😢

Comment (carissa-tang) [reply backend/models/user.model.ts:80]: i think that makes sense. so we delete the dependencies before the teacher. i can look into rolling back the changes for that too
